src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs

```diff
@@ -850,26 +851,51 @@
 
             // bucketize
             // prevent reallocation. it may not have 'count' entries, but it won't have more. 
-            var accumulator = new Dictionary<K, ArrayBuilder<T>>(items.Length, comparer);
+            //
+            // We store a mapping from keys to either a single item (very common in practice as this is used from
+            // callers that maps names to symbols with that name, and most names are unique), or an array builder of items.
+
+            var accumulator =  new Dictionary<K, object>(items.Length, comparer);
```

Comment:
>object [](http://example.com/codeflow?start=49&length=6)

Shouldn't this rather be `<K, (T, ArrayBuilder<T>?)?` to avoid boxing when `T` is a struct?

Reply:
there are currnetly only 3 callers in all of roslyn and it's all ref types.  i could put the `class` constraint here if that would be preferred.

Reply:
Probably worth doing, so we don't accidentally introduce a perf problem in the future.